### PR TITLE
stage 5 and 6

### PR DESCRIPTION
**Pull request by mohfez**
Stage 6 is complete but was not exhaustively tested.
Stage 5 is still in progress.

### CODE REVIEW

**Review by n8han11 — COMMENTED** [review]
By "any of the early entries being deleted", I mean that the previous commands are still being deleted as intended, but the number order (1-20) will remain intact.

File: Stages/stage_5_and_6.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#include "file_handling.h"
+
+// constants
+#define HIST_FILE_NAME ".hist_file"
+#define HISTORY_SIZE 20
+
+// Struct to represent a history entry
+typedef struct {
+    int number;
+    char command[256]; // Assuming a maximum command length of 255 characters
```
Comment (Deadly-Pigg) [inline Stages/stage_5_and_6.h]: 256 is not consistent with the shell limit; it should be 513. Use MAX_ARGS_QUANTITY provided.

File: shell.c [hunk 2]
```diff
@@ -33,13 +38,30 @@ int main(){
...
+
 		//While the command is a history invocation or alias then replace it 
 		//with the appropriate command from history or the aliased command 
 		//respectively (5 & 7)
+
+		// Check for history invocations and handle them
+        	char* history_invocation = invoke_from_history(args[0]);
+       		if (history_invocation != NULL) {
+            		// If it's a history invocation, replace the command with the history command
+            		free(*args);
```
Comment (Deadly-Pigg) [inline shell.c]: freeing *args too early breaks other commands (i.e, cd)
Thus, it is not needed.

File: shell.c [hunk 3]
```diff
@@ -33,13 +38,30 @@ int main(){
...
+            		*args = history_invocation;
```
Comment (Deadly-Pigg) [inline shell.c]: Only add to history if it's a non-history invoking command

**Comment by mohfez** [comment]
There was also another issue where it doesn't add the parameters to the history (i.e "cd banana" only adds "cd" to the history).

(possibly related to #31)

**Comment by mohfez** [comment]
+ issue about !-0 / equivalent (!0 ?)

**Comment by Deadly-Pigg** [comment]
> There was also another issue where it doesn't add the parameters to the history (i.e "cd banana" only adds "cd" to the history).
> 
> (possibly related to #31)

that seems to be a different problem compared to issue #31, since it's more so to do with cd imo. For history, you could probably save everything on the line which history saving is invoked on?

**Comment by Deadly-Pigg** [comment]
the branch is now broken and wont compile for some reason

**Comment by Deadly-Pigg** [comment]
but before (when it worked), the only issue that really arose was that there was no 'clear history' in the shell, so that must be implemented.

**Comment by VictorVictini** [comment]
lgtm

**Pull request merged**